### ISSUE TITLE
Occlusion between points

### ISSUE THREAD

**Issue by Vectornaut** [jsxgraph/jsxgraph#674]
As @Cleonis [notes](https://github.com/jsxgraph/jsxgraph/issues/633#issuecomment-2171743954), JSXGraph doesn't do any occlusion handling for 3d elements. Although general occlusion handling would be hard to implement, making points occlude each other correctly is relatively simple. I'd like to add this as a View3D option, because point–point occlusion can provide valuable depth cues even without more general occlusion.

My current prototype works by keeping a list of point elements alongside the associative array `View3D.objects` that contains all the view's elements. On every call to `update`, I select the visible points, sort them by depth, and then change their drawing order to match their depth order by calling `setLayer` in the SVG renderer. (*Edited to add: I use `setLayer` to move each point to the top of its current layer, not to move points to different layers. This avoids overriding users' layer choices. I don't think JSXGraph documentation makes any promises about how elements will be ordered within each layer, so I'm comfortable with changing that ordering.*)

**Comment by tom-berend** [comment]
Hi @Vectornaut.  Does that work?  SVG wasn't meant to have 'layers', it draws shapes in the order the SVG elements are presented in the HTML canvas.  Only HTML elements support zIndex.   Are you deleting and recreating the full collection of SVG points at each update?

As a proof-of-concept (mostly to see how fast it was), I was going to whack into the DOM and delete-sort-and-reinsert the SVG elements based on their z.  I was going to use the fill color as my key to tracking the points.

I built [this](https://communityreading.org/jsxgraph/3d.html) as a first step, and then got distracted by chasing down strange labels, which Alfred quickly identified and fixed.  Excited to hear you are ahead of me.

There is a good write-up on the problems with using 'layers' in SVG [here](https://zzz.dog/extras).

**Comment by Vectornaut** [comment]
@tom-berend: To clarify, I use `setLayer` to re-order the points within each layer, not to move points to different layers. I've updated the issue to reflect this.

You can see my prototype code on [this branch](https://github.com/Vectornaut/jsxgraph/tree/depth-ordering-points). I've been testing it on the `intersection_sphere_sphere.html` and `polygon3d.html` examples. Thanks for sharing the prototype you built! If there's a way for me to see the non-bundled source files, I could look at your implementation.

>  Does that work? SVG wasn't meant to have 'layers', it draws shapes in the order the SVG elements are presented in the HTML canvas. Only HTML elements support zIndex.

Here's how the SVG renderer's layer system works, as far as I can tell. The layers are a [sequence of groups](https://github.com/jsxgraph/jsxgraph/blob/6d1fd94ea6969fe6a2e4264dca3b0a75eb4022f1/src/renderer/svg.js#L250-L260)—twenty, by default—which are direct children of the SVG root. Every element's SVG rendering node belongs to one of these groups. By default, elements are [sorted](https://github.com/jsxgraph/jsxgraph/blob/6d1fd94ea6969fe6a2e4264dca3b0a75eb4022f1/src/options.js#L3467-L3526) into the layer groups based on their type. This is why, for example, text is drawn over points and points are drawn over lines by default.

The `SVGRenderer.setLayer` method can be used to change an element's layer. Because it uses `appendChild` to move SVG nodes, it has the side effect of sending the target element to the top of the layer you put it in—even if you put it in the layer it already belongs to.

> Are you deleting and recreating the full collection of SVG points at each update?

When [`appendChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild) is used to append a node that's already part of the document, it "moves [the node] from its current position to the new position," which I believe is the end of the child list of the node you're appending to. I don't know how browsers typically implement it, so I guess it's possible for it to work by deleting and recreating the node in some sense, but I get the impression that people think of it as moving the node.

**Comment by alfredwassermann** [comment]
@Vectornaut's plan might work! The way to go is to reorder nodes in an SVG layer. Deleting and recreating is probably too slow, but reordering might work. Actually, JSXGraph is doing a similar action with the attribute `dragToTopOfLayer`, see https://jsxgraph.org/docs/symbols/JXG.GeometryElement.html#dragToTopOfLayer. 
For redrawing, browsers seem to take out the SVG graphic, repaint it and show it again.

**Comment by Vectornaut** [comment]
> Actually, JSXGraph is doing a similar action with the attribute `dragToTopOfLayer`

This is good to know! Maybe it's worth adding a little `moveToTopOfLayer` utility method to `Board`. The method could be used in both places.

@alfredwassermann, do you have any thoughts on:
- How to name the option that would turn point occlusion on?
- How to check whether point occlusion is slowing down the interface too much?

**Comment by tom-berend** [comment]
@Vectornaut  I have replaced the TS syntax with plain JSXGraph [here](https://communityreading.org/jsxgraph/3d.html).  It was a reminder of how twitchy the JSXGraph syntax is without an editor to do the gruntwork.

There's another rendering trick that might be particularly fast.  The [use](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/use) element duplicates another SVG element by reference.   Maybe hide the 'real' points and point to them, updating the entire pointer list on every update.  I'm thinking that if the browser has already parsed and rendered the origintal points into a hidden div,  it might be very fast to redraw them.

Excited by this project.  Won't be able to download your branch until Monday.

**Comment by Vectornaut** [comment]
> I have replaced the TS syntax with plain JSXGraph [here](https://communityreading.org/jsxgraph/3d.html). It was a reminder of how twitchy the JSXGraph syntax is without an editor to do the gruntwork.

@tom-berend: I see—you're using the current release of JSXGraph. (I'd assumed you were using a modified version.) Does this example code currently do occlusion handling, or is it a testbed that doesn't include any occlusion handling yet?

> There's another rendering trick that might be particularly fast. The [use](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/use) element duplicates another SVG element by reference.

I haven't tried that trick yet, because I think it would require deeper changes to the SVG renderer, but I agree that it will be worth a try if re-ordering SVG nodes is too slow. That's one reason I [asked](https://github.com/jsxgraph/jsxgraph/issues/674#issuecomment-2183407920) how to check whether the current point occlusion routine is slowing down the interface too much. A good benchmark will help us decide whether to invest in this deeper change.

> Excited by this project. Won't be able to download your branch until Monday.

If all goes well, I may have a pull request submitted by Monday. If so, I'd appreciate your comments on it. If we want to try a method substantially different from the one I'm using, one of us could always submit a second PR.

**Comment by tom-berend** [comment]
@Vectornaut  No occlusion handling yet, but the intention was a proof-of-concept (thank you Cleonis).  I hoped that whacking into the DOM wouldn't be too hard.  

Occlusion would amazing with the new 3D rendering.  JSXGraph would become a wicked fun tool for writing engineering simulations and simple games. 

Looking forward to trying your PR.

**Comment by tom-berend** [comment]
@Vectornaut  I played a bit more with points [here](https://communityreading.org/jsxgraph/box3d.html), and also with the new Polygon3D element. 

Would you consider adding Polygon3D elements to your zIndex PR?  Perhaps the 'point' for calculation would be the centroid of the polygon.  That would allow us to create and manipulate 3D geometric solids with flat sides, and to approximate solids with curved sides.

Looking forward to trying your PR.

**Comment by Vectornaut** [comment]
@tom-berend I just posted the pull request! It's #676.

> Would you consider adding Polygon3D elements to your zIndex PR?

I'm impressed with how well the simple depth-ordering in Zdog works, as discussed in the documentation you linked to [earlier](https://github.com/jsxgraph/jsxgraph/issues/674#issuecomment-2182588724), so I'd be interested to try an expanded depth ordering system in JSXGraph. Although I want to keep pull request #676 focused on points, I think the implementation could be extended to other 3d primitives.

The tricky part of generalizing the depth-sorting system would be deciding how to deal with layers. Objects like points and polygons are placed on different layers by default, so depth-ordering them would mean either implementing cross-layer depth-sorting or requiring the user to manually put depth-sorted objects in the same layer.

**Comment by alfredwassermann** [comment]
We could experiment with putting all 3D elements (respect. their 2D images) beside labels into one dedicated layer.
For wire frame elements we have to change the mesh from being a single SVG path to be an array of paths. Of course, the price will be a performance penalty.

**Comment by tom-berend** [comment]
@Vectornaut  Not working for me.  [Here's my demo](https://communityreading.org/jsxgraph/box3d.html).  I tried with both the 'develop' branch and the 'Vectornaut-spheres' branches.  (the demo is using 'develop').   I have added the 
 attribute 'depthOrderPoints:true' to both the Point3D and the board.   Need some help.
 
 @alfredwassermann I vote for the performance penalty.  Machines are getting faster, and content creators are getting slower.  Turn occlusion ON by default.

**Comment by Vectornaut** [comment]
### Option usage

@tom-berend: `depthOrderPoints` is a View3D option, so it shouldn't be set for any element other than a View3D. If you're on the `develop` branch (commit 3c3b28f or later), you can see usage examples in `polygon3d.html`, `intersection_sphere_sphere.html`, and the other example files mentioned in #676.

### Patching your example
Switching to the following version of `box3d.js`, on commit 5511551 of JSXGraph, makes your example work for me!
```
import { TXG } from "./tsxgraph.js";
// can't easily call a TS function or class from JS.  so we call as we load this script.
box3d();
function box3d() {
    const TSX = TXG.TSXGraph.initBoard('jxgbox');
    let bound = [-5, 5];
    let view = TSX.view3D(-6, -3, 8, 8, bound, bound, bound, { depthorderpoints: true });
    let PI = Math.PI;
    TSX.text([-3.5, 4.85], 'Rotation');
    let rxSlide = TSX.slider([-4.5, 4.6], [-2.5, 4.6], [-PI, 0, PI], { name: '' });
    let rySlide = TSX.slider([-4.5, 4.3], [-2.5, 4.3], [-PI, 0, PI], { name: '' });
    let rzSlide = TSX.slider([-4.5, 4.0], [-2.5, 4.0], [-PI, 0, PI], { name: '' });
    TSX.text([-0.5, 4.85], 'Translation');
    let txSlide = TSX.slider([-1.5, 4.6], [.5, 4.6], [-3, 0, 3], { name: '' });
    let tySlide = TSX.slider([-1.5, 4.3], [.5, 4.3], [-3, 0, 3], { name: '' });
    let tzSlide = TSX.slider([-1.5, 4.0], [.5, 4.0], [-3, 0, 3], { name: '' });
    TSX.text([2, 4.85], 'Scale');
    let sxSlide = TSX.slider([1.5, 4.6], [3.5, 4.6], [.1, 1, 3], { name: '' });
    let sySlide = TSX.slider([1.5, 4.3], [3.5, 4.3], [.1, 1, 3], { name: '' });
    let szSlide = TSX.slider([1.5, 4.0], [3.5, 4.0], [.1, 1, 3], { name: '' });
    let hex = (n) => n.toString(16).padStart(2, '0');
    let points = [];
    let low = -2;
    let high = 2;
    for (let i = low; i <= high; i++) {
        for (let j = low; j <= high; j++) {
            for (let k = low; k <= high; k++) {
                let color = '#' + hex((i + 3) * 40) + hex((j + 3) * 40) + hex((k + 3) * 40);
                let color2 = '#' + hex((i + 2) * 40) + hex((j + 2) * 40) + hex((k + 2) * 40);
                points.push({
                    point: view.point3D([i, j, k], { gradient: 'linear', fillColor: color, gradientSecondColor: color2, size: 20, withLabel: false }),
                    x: i,
                    y: j,
                    z: k,
                    xOrig: i,
                    yOrig: j,
                    zOrig: k,
                });
            }
        }
    }
    // https://www.tutorialspoint.com/computer_graphics/3d_transformation.htm
    let transform = () => {
        // rotation around x
        let cosX = Math.cos(rxSlide.Value());
        let sinX = Math.sin(rxSlide.Value());
        let rotX = [[1, 0, 0, 0],
            [0, cosX, -sinX, 0],
            [0, sinX, cosX, 0],
            [0, 0, 0, 1]];
        // rotation around y
        let cosY = Math.cos(rySlide.Value());
        let sinY = Math.sin(rySlide.Value());
        let rotY = [[cosY, 0, sinY, 0],
            [0, 1, 0, 0],
            [-sinY, 0, cosY, 0],
            [0, 0, 0, 1]];
        // rotation around z
        let cosZ = Math.cos(rzSlide.Value());
        let sinZ = Math.sin(rzSlide.Value());
        let rotZ = [[cosZ, -sinZ, 0, 0],
            [sinZ, cosZ, 0, 0],
            [0, 0, 1, 0],
            [0, 0, 0, 1]
        ];
        points.map((pointN) => {
            // pick up translate values
            let px = txSlide.Value();
            let py = tySlide.Value();
            let pz = tzSlide.Value();
            // pick up scale values
            let sx = sxSlide.Value();
            let sy = sySlide.Value();
            let sz = szSlide.Value();
            let point = [pointN.xOrig * sx, pointN.yOrig * sy, pointN.zOrig * sz, 0];
            point = TSX.MatrixMath.matVecMult(rotX, point);
            point = TSX.MatrixMath.matVecMult(rotY, point);
            point = TSX.MatrixMath.matVecMult(rotZ, point);
            pointN.point.setPosition([px + point[0], py + point[1], pz + point[2]]);
        });
    };
    rxSlide.onChange(() => transform());
    rySlide.onChange(() => transform());
    rzSlide.onChange(() => transform());
    txSlide.onChange(() => transform());
    tySlide.onChange(() => transform());
    tzSlide.onChange(() => transform());
    sxSlide.onChange(() => transform());
    sySlide.onChange(() => transform());
    szSlide.onChange(() => transform());
}
```

**Comment by Vectornaut** [comment]
> For wire frame elements we have to change the mesh from being a single SVG path to be an array of paths.

Could you explain why that change would be needed? I have the impression that Zdog depth-sorts each element—for example, each mesh—based on the depth of a single point. It seems to me that JSXGraph could do the same without splitting up the SVG path used to render a mesh.

**Comment by tom-berend** [comment]
@Vectornaut  AMAZING.   I have increased my [demo](https://communityreading.org/jsxgraph/box3d.html) to 1,000 points, and still perfectly responsive.  Both to mouse and slider interactions.  Including on my phone.  WOW.

Of course, now I want to add a physics engine and watch them bounce around as the cube collapses. This is FUN. 

Thank you !!!

**Comment by tom-berend** [comment]
@alfredwassermann  The 3D wireframes are delightful.  They are constructions that can be easily created by users if @Vectornaut adds zIndex capabilities to polygon3D. 

Demonstrating or culling hidden faces requires polygons to indicate which side is front or back, and allowing different colors, opacity, and line attributes for each side.  This might be a useful addition to polygon3D.

Calculating diffuse and specular lighting calculations only requires exposing where the camera is.   The lights themselves are just part of the user's construction.  Again, perhaps a useful addition.

'Sphere' and other complex shapes built from polygons would behave as they do in WebGL.   But never as perfectly, the z-fighting explanation in (zDog)[https://zzz.dog/extras] will always be there.  WebGL will occlude parts of a triangle, SVG is all or nothing.

**Comment by Vectornaut** [comment]
This feature has now been implemented, in pull request #676. Thanks for the great discussion!
